App/Services/UriHandler.cs

```diff
@@ -0,0 +1,150 @@
+
+public interface IUriHandler : IAsyncDisposable
+{
+    public Task HandleUri(Uri uri, CancellationToken ct = default);
+}
+
+public class UriHandler(
+    ILogger<UriHandler> logger,
+    IRpcController rpcController,
+    IUserNotifier userNotifier,
+    IRdpConnector rdpConnector) : IUriHandler
+{
+    private const string OpenWorkspacePrefix = "/v0/open/ws/";
+
+    internal class UriException(string title, string detail) : Exception
+    {
+        internal readonly string Title = title;
+        internal readonly string Detail = detail;
+    }
+
+    public async Task HandleUri(Uri uri, CancellationToken ct = default)
+    {
+        try
+        {
+            await HandleUriThrowingErrors(uri, ct);
+        }
+        catch (UriException e)
+        {
+            await userNotifier.ShowErrorNotification(e.Title, e.Detail, ct);
+        }
+    }
+
+    private async Task HandleUriThrowingErrors(Uri uri, CancellationToken ct = default)
+    {
+        if (uri.AbsolutePath.StartsWith(OpenWorkspacePrefix))
+        {
+            await HandleOpenWorkspaceApp(uri, ct);
+            return;
+        }
+
+        logger.LogWarning("unhandled URI path {path}", uri.AbsolutePath);
+        throw new UriException("URI handling error",
+            $"URI with path {uri.AbsolutePath} is unsupported or malformed");
+    }
+
+    public async Task HandleOpenWorkspaceApp(Uri uri, CancellationToken ct = default)
+    {
+        const string errTitle = "Open Workspace Application Error";
+        var subpath = uri.AbsolutePath[OpenWorkspacePrefix.Length..];
+        var components = subpath.Split("/");
+        if (components.Length != 4 || components[1] != "agent")
+        {
+            logger.LogWarning("unsupported open workspace app format in URI {path}", uri.AbsolutePath);
+            throw new UriException(errTitle, $"Failed to open {uri.AbsolutePath} because the format is unsupported.");
+        }
+
+        var workspaceName = components[0];
+        var agentName = components[2];
+        var appName = components[3];
+
+        var state = rpcController.GetState();
+        if (state.VpnLifecycle != VpnLifecycle.Started)
+        {
+            logger.LogDebug("got URI to open workspace {workspace}, but Coder Connect is not started", workspaceName);
+            throw new UriException(errTitle,
+                $"Failed to open application on {workspaceName} because Coder Connect is not started.");
+        }
+
+        Workspace workspace;
+        try
+        {
+            workspace = state.Workspaces.Single(w => w.Name == workspaceName);
+        }
+        catch (InvalidOperationException) // Single() throws this when nothing matches.
+        {
+            logger.LogDebug("got URI to open workspace {workspace}, but the workspace doesn't exist", workspaceName);
+            throw new UriException(errTitle,
+                $"Failed to open application on workspace {workspaceName} because it doesn't exist");
+        }
```

Comment:
A nicer pattern I've used that doesn't require a try/catch and predeclaring the variable outside the try block is:
```cs
var workspace = state.Workspaces.FirstOrDefault(w => w.Name == workspaceName);
if (workspace is null)
{
  // ...
  throw new Exception("");
}
```